gcsfs/extended_gcsfs.py

```diff
@@ -0,0 +1,228 @@
+    to the parent class GCSFileSystem for default processing.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.grpc_client = None
+        self.storage_control_client = None
+        # initializing grpc and storage control client for Hierarchical and
+        # zonal bucket operations
+        self.grpc_client = asyn.sync(self.loop, self._create_grpc_client)
+        self._storage_control_client = asyn.sync(
+            self.loop, self._create_control_plane_client
+        )
+        self._storage_layout_cache = {}
+
+    async def _create_grpc_client(self):
+        if self.grpc_client is None:
+            return AsyncGrpcClient(
+                client_info=ClientInfo(user_agent=f"{USER_AGENT}/{version}"),
+            ).grpc_client
+        else:
+            return self.grpc_client
+
+    async def _create_control_plane_client(self):
+        # Initialize the storage control plane client for bucket
+        # metadata operations
+        client_info = gapic_v1.client_info.ClientInfo(
+            user_agent=f"{USER_AGENT}/{version}"
+        )
+        return storage_control_v2.StorageControlAsyncClient(
+            credentials=self.credentials.credentials, client_info=client_info
+        )
+
+    async def _get_bucket_type(self, bucket):
+        if bucket in self._storage_layout_cache:
+            return self._storage_layout_cache[bucket]
+        try:
+
+            # Bucket name details
+            bucket_name_value = f"projects/_/buckets/{bucket}/storageLayout"
+            # Make the request to get bucket type
+            response = await self._storage_control_client.get_storage_layout(
+                name=bucket_name_value
+            )
+
+            if response.location_type == "zone":
+                return BucketType.ZONAL_HIERARCHICAL
+            else:
+                # This should be updated to include HNS in the future
+                return BucketType.NON_HIERARCHICAL
+        except api_exceptions.NotFound:
+            print(f"Error: Bucket {bucket} not found or you lack permissions.")
+            return BucketType.UNKNOWN
+        except Exception as e:
+            logger.error(
+                f"Could not determine bucket type for bucket name {bucket}: {e}"
+            )
+            # Default to UNKNOWN
+            return BucketType.UNKNOWN
+
+    _sync_get_bucket_type = asyn.sync_wrapper(_get_bucket_type)
+
+    def _open(
+        self,
+        path,
+        mode="rb",
+        block_size=None,
+        cache_options=None,
+        acl=None,
+        consistency=None,
+        metadata=None,
+        autocommit=True,
+        fixed_key_metadata=None,
+        generation=None,
+        **kwargs,
+    ):
+        """
+        Open a file.
+        """
+        bucket, _, _ = self.split_path(path)
+        bucket_type = self._sync_get_bucket_type(bucket)
+        self._storage_layout_cache[bucket] = bucket_type
```

Comment:
Storing in the cache should happen in the _sync_get_bucket_type method, not here